components/WeekNavigator.tsx

```diff
@@ -1,30 +1,74 @@
 import { Button, Stack, Typography } from "@mui/material";
-import React from "react";
+import React, { Dispatch, SetStateAction, useState } from "react";
+import { SitSchedule } from "../types/sitTypes";
+import { DateTime } from "luxon";
 
 export default function WeekNavigator({
-    weekNumber,
-    weekOffset,
-    loadingPreviousWeek,
-    loadingNextWeek,
-    onUpdateWeekOffset,
+    initialCachedSchedules,
+    setCurrentSchedule,
 }: {
-    weekNumber: number;
-    weekOffset: number;
-    loadingPreviousWeek: boolean;
-    loadingNextWeek: boolean;
-    // eslint-disable-next-line no-unused-vars
-    onUpdateWeekOffset: (modifier: number) => void;
+    initialCachedSchedules: { [weekOffset: number]: SitSchedule };
+    setCurrentSchedule: Dispatch<SetStateAction<SitSchedule>>;
 }) {
+    const [weekOffset, setWeekOffset] = useState(0);
+    const [loadingNextWeek, setLoadingNextWeek] = useState(false);
+    const [loadingPreviousWeek, setLoadingPreviousWeek] = useState(false);
+    const [cachedSchedules, setCachedSchedules] = useState<{ [weekOffset: number]: SitSchedule }>(
+        initialCachedSchedules
+    );
+    async function updateWeekOffset(modifier: number) {
+        switch (modifier) {
+            case -1:
+                setLoadingPreviousWeek(true);
+                break;
+            case 1:
+                setLoadingNextWeek(true);
+                break;
+        }
+        const currentWeekOffset = modifier === 0 ? 0 : weekOffset + modifier;
+        let cachedSchedule = cachedSchedules[currentWeekOffset];
+        if (cachedSchedule === undefined) {
+            cachedSchedule = await fetch("api/schedule", {
+                method: "POST",
+                body: JSON.stringify({ weekOffset: currentWeekOffset }),
+            }).then((r) => r.json());
+            if (cachedSchedule === undefined) {
+                setLoadingPreviousWeek(false);
+                setLoadingNextWeek(false);
+                throw new Error("Failed to fetch schedule");
+            }
+            setCachedSchedules({ ...cachedSchedules, [currentWeekOffset]: cachedSchedule });
+        }
+        setWeekOffset(currentWeekOffset);
+        setCurrentSchedule(cachedSchedule);
+        setLoadingPreviousWeek(false);
+        setLoadingNextWeek(false);
+        // Pre-fetch next schedule (in same direction) if not in cache
+        const nextWeekOffset = currentWeekOffset + modifier;
+        if (nextWeekOffset in cachedSchedules) {
+            return;
+        }
+        const nextSchedule = await fetch("api/schedule", {
+            method: "POST",
+            body: JSON.stringify({ weekOffset: nextWeekOffset }),
+        }).then((r) => r.json());
+        if (nextSchedule != undefined) {
+            setCachedSchedules({ ...cachedSchedules, [nextWeekOffset]: nextSchedule });
+        }
+    }
+
+    const weekNumber = DateTime.fromISO(initialCachedSchedules[weekOffset]!.days[0]!.date).weekNumber;
```

Comment:
This occurs when navigating two weeks back

![image](https://github.com/mathiazom/sit-rezervo-confgen/assets/24361490/b8811abb-ea31-4e7a-8775-264792966d4a)

Reply:
Introduced in 3f631f259699cd9edc392bc2036090e353f3b074

Reply:
Fixed in https://github.com/mathiazom/sit-rezervo-confgen/pull/29/commits/e89794db6f03dd898c48354bbfb03c9c705f6fb2

Let me know if you want me to rebase and amend this fix to the original commit (https://github.com/mathiazom/sit-rezervo-confgen/commit/3f631f259699cd9edc392bc2036090e353f3b074)